umbraco/Umbraco-CMS
# Get MB with PureLive working for netcore

Gets ModelsBuilder in all modes working

* PureLive - for dynamically recompiling DLLs + Views
* Live App_Data - for dynamically re-generating csharp models
* App_Data - for regenerating csharp models on demand

## Completed

* PureLiveModelFactory:
  * We cannot use in-memory / non-file assemblies for pure live since this is not compatible with the default razor view engine and there are no public APIs to make that possible without basically copying the entire csharp for the view engine code (a lot). Instead we use temp DLL files. For the razor view engine to work with other assemblies they must have a physical file path.
  * Ensure generated DLLs are added (and removed) from the ApplicationParts since this is how the razor view engine loads in external assemblies.
  * New event ModelsChanged for when they are changed, this is used by other code to 'refresh' the razor view engine. Cannot use event aggregator since it cannot be async.
  * Ensure UmbracoAssemblyLoadContext is used in conjunction with `AssemblyLoadContext.Default.Resolving` event to tell the default load context how to find custom assemblies loaded with the custom context
* Razor Engine
  * There's a huge block of research notes put into UmbracoBuilderExtensions explaining how we go about hot reloading the razor engine when assemblies change. This is done with the `RefreshingRazorViewEngine` (see notes)
* ModelsBuilderSettings
  * Enabled no longer exists. This didn't make any sense since you cannot "Disable" models builder. All this was doing was configuring some events and UI. Instead we just rely on the ModelsMode to adjust how things work. To 'disable' strongly typed models generation and just use the the base IPublishedContent models, just use ModelsMode = "Nothing"
  * EnableFactory no longer exists. This didn't make any sense at all. If you disable this, absolutely nothing works because you cannot have no MB factory. It is responsible for creating non-typed models (IPublishedContent) and strongly typed models if the ModelsMode is configured for that.
  * PureLive is the default in code - since it's the default in config and what we ship with.
* Models builder components/composers to UmbracoBuilder ext methods
* Fixes up ModelsBuilder Dashboard
 * Removes system wide mutex and replaces by using MainDom - system wide mutex wont work anyways on Azure or Linux (so now netcore compat on linux)

## Others:

* Simplifies IMainDom interface
* Change event on ServerVariablesParser to event aggregator
* Remove old unused/obsolete code
* Missing null checks in PublishedSnapshotService - these are in v8 but the checks didn't carry over during migration to netcore, so they are fixed again

## Testing

* Have PureLive mode configured (it is the default)
* Change one of your views to be strongly typed by doing this in your razor view: `@inherits Umbraco.Web.Common.AspNetCore.UmbracoViewPage<Umbraco.Web.PublishedModels.Hello>` instead of just `@inherits Umbraco.Web.Common.AspNetCore.UmbracoViewPage` ... where `Hello` is the PascalCase of your doc type alias
* Also add this to your razor file and output the value of this in your view. It will give you the assembly qualified name of the model being used on this page: 
   ```cs
    var modelType = (this.GetType().BaseType.GenericTypeArguments.Length > 0
       ? this.GetType().BaseType.GenericTypeArguments[0]
       : this.GetType().BaseType.BaseType.GenericTypeArguments[0]).AssemblyQualifiedName;
   ```
  * This should all just work
  * Now go modify that document type - easiest thing is to just change the name (not alias) of a property. This will regenerate all models and also refresh the razor engine
  * Go to the front-end, this will all 'just work' and you will see that your assembly version of your model has changed
  * Go a step further and add this to your razor output: `@Model.Blah` (assuming you don't have a `blah` property type yet). Render the page and you'll get an error because that model doesn't contain that property. 
  * Go add this property type as a textstring to your document type
  * Go to the front-end, it will 'just work' and you will see that your assembly version of your model has changed
* Change to LiveAppData
  * This will re-generate the models as .cs class in the `\Umbraco\models` folder, one for each document type and will re-generate them each time you change a doc type - you can test this by just looking at the files there and changing doc types
* Change to AppData
  * This will re-generate the models as .cs class in the `\Umbraco\models` folder, one for each document type and will re-generate them on demand - you can test this by going to the MB dashboard and clicking re-generate. Note that this option will not be there for the PureLive mode, you can test that too!

## TODO: 

* [ ] Benchmark testing to ensure there are no mem leaks == will create a separate task.
